### PR TITLE
HDDS-10559. Add a warning or a check to run repair tool as System user

### PR DESCRIPTION
**Pull request by DaveTeng0**
## What changes were proposed in this pull request?
We added a repair tool for snapshot chain fix as part of PR: https://github.com/apache/ozone/pull/6386

This check was missed in the original PR. This task is to add a check or warning to make sure that the repair tool is run as system user (default to hdfs)

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-10559

## How was this patch tested?
Unit test

### CODE REVIEW

**Comment by DaveTeng0** [comment]
cc. @hemantk-12 @errose28

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java [hunk 1]
```diff
@@ -61,4 +71,33 @@ public OzoneConfiguration getOzoneConf() {
...
+    if (!shouldProceed) {
+      System.out.println("Aborting command.");
+      return 1;
+    }
+    System.out.println("Run as user: " + currentUser);
+
+    TracingUtil.initTracing("shell", createOzoneConfiguration());
+    String spanName = "ozone repair " + String.join(" ", argv);
+    return TracingUtil.executeInNewSpan(spanName,
+        () -> super.execute(argv));
```
Comment (errose28) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:92]: There's no network calls going out as part of repair commands. They run locally when nodes are stopped, so we don't need request tracing here like have for other commands.

File: hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/repair/TestOzoneRepair.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    // restore system streams
+    System.setOut(OLD_OUT);
+    System.setErr(OLD_ERR);
+  }
+
+  @Test
+  void testOzoneRepairWhenUserIsOzoneDefaultUser() throws Exception {
+    OzoneRepair ozoneRepair = spy(new OzoneRepair());
+    doReturn(OZONE_DEFAULT_USER).when(ozoneRepair).getSystemUserName();
```
Comment (errose28) [inline hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/repair/TestOzoneRepair.java:66]: I think you can use `System#setProperty` to alter the perceived user name without needing `spy`.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java [hunk 3]
```diff
@@ -22,8 +22,11 @@
 import org.apache.hadoop.hdds.cli.GenericCli;
 import org.apache.hadoop.hdds.cli.HddsVersionProvider;
 import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.hdds.tracing.TracingUtil;
 import picocli.CommandLine;
 
+import static org.apache.hadoop.fs.ozone.Constants.OZONE_DEFAULT_USER;
```
Comment (errose28) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:28]: This is just used to create a default user directory in the Ozone filesystem. It is not related to the user the OM or any other component is run as, although "hdfs" does happen to be a common choice for that.

**Review by errose28 — COMMENTED** [review]
Thanks for adding this @DaveTeng0. The repair command could have uses on all nodes, not just OM. I think it is also difficult to figure out which user is usually being used to run each process from this command. I'm thinking that instead of trying to check the value against a "correct" value that's hard to determine, we can just print a message to stderr every time like:
```
Running as user <user>. Make sure this is the same user used to run the Ozone process.
```
The repair command will be infrequently used, so I think printing messages like this every time are ok. If someone knows a reliable way to get the user that runs each process though let me know and we can use that instead.

**Comment by DaveTeng0** [comment]
> Thanks for adding this @DaveTeng0. The repair command could have uses on all nodes, not just OM. I think it is also difficult to figure out which user is usually being used to run each process from this command. I'm thinking that instead of trying to check the value against a "correct" value that's hard to determine, we can just print a message to stderr every time like:
> 
> ```
> Running as user <user>. Make sure this is the same user used to run the Ozone process.
> ```
> 
> The repair command will be infrequently used, so I think printing messages like this every time are ok. If someone knows a reliable way to get the user that runs each process though let me know and we can use that instead.

Yeah it makes sense! updated command message accordingly! Thanks @errose28 !

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java (hunk 1 above)
Comment (DaveTeng0) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:92]: Yeah makes sense! removed tracing.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java (hunk 3 above)
Comment (DaveTeng0) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:28]: makes sense! removed!

File: hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/repair/TestOzoneRepair.java (hunk 2 above)
Comment (DaveTeng0) [reply hadoop-ozone/tools/src/test/java/org/apache/hadoop/ozone/repair/TestOzoneRepair.java:66]: yeah makes sense! Changed to use System#setProperty instead.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java [hunk 4]
```diff
@@ -61,4 +66,27 @@ public OzoneConfiguration getOzoneConf() {
...
+
+    return super.execute(argv);
+  }
+
+  public  String getSystemUserName() {
+    return System.getProperty("user.name");
+  }
+
+  public  String getConsoleReadLineWithFormat(String currentUser) {
+    return System.console().readLine(String.format(WARNING_SYS_USER_MESSAGE, currentUser));
```
Comment (errose28) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:89]: Looks like other commands like [DecomissionSubCommand](https://github.com/apache/ozone/blob/92bc617801e31c3ad8a75602e1167300c6936b33/hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DecommissionSubCommand.java#L63) are using `Scanner`. We can use that here too for consistency. `Scanner` also supports redirecting stdin like [this](https://github.com/apache/ozone/blob/92bc617801e31c3ad8a75602e1167300c6936b33/hadoop-hdds/tools/src/test/java/org/apache/hadoop/hdds/scm/cli/datanode/TestDecommissionSubCommand.java#L79) so no need for `Spy` in the test when using this approach.

Comment (errose28) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:89]: Also make sure the prompt goes to stderr. There may be repair commands were the output is redirected for reference later, and we don't want the prompt to get lost there.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java [hunk 5]
```diff
@@ -61,4 +66,27 @@ public OzoneConfiguration getOzoneConf() {
   public static void main(String[] argv) throws Exception {
     new OzoneRepair().run(argv);
   }
+
+  @Override
+  public int execute(String[] argv) {
+    String currentUser = getSystemUserName();
+    String s = getConsoleReadLineWithFormat(currentUser);
+    boolean shouldProceed = Boolean.parseBoolean(s) || "y".equalsIgnoreCase(s);
```
Comment (errose28) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:74]:
```suggestion
    boolean shouldProceed = "y".equalsIgnoreCase(s);
```

I think this is sufficient. I'm not sure we need to support people typing "true" into the prompt.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java (hunk 4 above)
Comment (errose28) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:89]: In the test we can then make sure that this stderr message also contains the user name. Currently we are only checking the "run as user" message.

**Review by errose28 — COMMENTED** [review]
Thanks for updating @DaveTeng0

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java (hunk 5 above)
Comment (DaveTeng0) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java:74]: yeah, 'y' should be good! updated.

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java (hunk 4 above)
Comment (DaveTeng0) [reply hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/repair/OzoneRepair.java]: make sense! updated! Thanks @errose28!

**Comment by DaveTeng0** [comment]
Hello! If no further new comments, please feel free to merge! Thanks!

**Review by hemantk-12 — APPROVED** [review]
Thanks, @DaveTeng0 for the change.

Overall looks good to me.

**Comment by hemantk-12** [comment]
Thanks, @DaveTeng0 for the change and @errose28 for the review.

**Pull request merged**